File: src/light/light.cpp
#include "light.h"

void CLight::LightGet(byte ID, int* NumVal, float* val, char** unit) 
{
	if (conf == false)
	{
		chmc5883.begin();
	    mcp3428_1.init(MCP342X::L, MCP342X::L);
	    mcp3428_2.init(MCP342X::L, MCP342X::F);
		conf = true;
	}

	if (ID == 0x10)
		HMC5883L(NumVal, val, unit);
	else if (ID == 0x11)
		HIH6130(NumVal, val, unit);
	else if (ID == 0x12)
		APDS9006(NumVal, val, unit);
	else if (ID == 0x13)
		TSL260RD(NumVal, val, unit);
	else if (ID == 0x14)
		TSL250RD(NumVal, val, unit);
	else if (ID == 0x15)
		MLX75305(NumVal, val, unit);
	else if (ID == 0x16)
		ML8511(NumVal, val, unit);
	else if (ID == 0x17)
		TMP421(NumVal, val, unit);

	else
		NumVal = 0;
}

void CLight::HMC5883L(int* NumVal, float* val, char** unit)
{
	*NumVal = 3;
	unit[0] = "Gx";
	unit[1] = "Gy";
	unit[2] = "Gz";

    chmc5883.getEvent(&event);

    val[0] = event.magnetic.x;
    val[1] = event.magnetic.y;
    val[2] = event.magnetic.z;
}

void CLight::HIH6130(int* NumVal, float* val, char** unit)
{
	*NumVal = 2;
	unit[0] = "C";
	unit[1] = "RH";

	chih6.HIH_fetch_humidity_temperature(val);
}

void CLight::APDS9006(int* NumVal, float* val, char** unit)
{
	*NumVal = 1;
	unit[0] = "AL Raw";

	mcp3428_2.selectChannel(MCP342X::CHANNEL_0, MCP342X::GAIN_1);
	val[0] = mcp3428_2.readADC();
}

void CLight::TSL260RD(int* NumVal, float* val, char** unit)
{
	*NumVal = 1;
	unit[0] = "IR Raw";

	mcp3428_1.selectChannel(MCP342X::CHANNEL_1, MCP342X::GAIN_1);
	val[0] = mcp3428_1.readADC();
}

void CLight::TSL250RD(int* NumVal, float* val, char** unit)
{
	*NumVal = 1;
	unit[0] = "VL Raw";
	
	mcp3428_1.selectChannel(MCP342X::CHANNEL_3, MCP342X::GAIN_1);
	val[0] = mcp3428_1.readADC();
}

void CLight::MLX75305(int* NumVal, float* val, char** unit)
{
	*NumVal = 1;
	unit[0] = "AL Raw";

	mcp3428_1.selectChannel(MCP342X::CHANNEL_0, MCP342X::GAIN_1);
	val[0] = mcp3428_1.readADC();
}

void CLight::ML8511(int* NumVal, float* val, char** unit)
{
	*NumVal = 1;
	unit[0] = "UV Raw";

	mcp3428_1.selectChannel(MCP342X::CHANNEL_2, MCP342X::GAIN_1);
	val[0] = mcp3428_1.readADC();
}

void CLight::TMP421(int* NumVal, float* val, char** unit)
{
	*NumVal = 1;
	unit[0] = "C";

    val[0] = ctmp421.GetTemperature();
}

File: talker/light.py
import math

class Lightsense():

	def lightDecode(self, ID, textlist):
		# Call function according to the sensor ID
		light_dict = {}
		if ID == 0x10:
			light_dict['HMC5883'] = self.decodeHMC5883(textlist)
			return light_dict
		elif ID == 0x11:
			light_dict['HIH6130'] = self.decodeHIH6130(textlist)
			return light_dict
		elif ID == 0x12:
			light_dict['APDS9006'] = self.decodeAPDS9006(textlist)
			return light_dict
		elif ID == 0x13:
			light_dict['TSL260RD'] = self.decodeTSL260RD(textlist)
			return light_dict
		elif ID == 0x14:
			light_dict['TSL250RD'] = self.decodeTSL250RD(textlist)
			return light_dict
		elif ID == 0x15:
			light_dict['MLX75305'] = self.decodeMLX75305(textlist)
			return light_dict
		elif ID == 0x16:
			light_dict['ML8511'] = self.decodeML8511(textlist)
			return light_dict
		elif ID == 0x17:
			light_dict['TMP421'] = self.decodeTMP421(textlist)
			return light_dict

		else:
			return 0

	def decodeHMC5883(self, textlist):
		# Magnetic field
		# Gx
		val1 = int(textlist[0], 16) << 8
		val2 = int(textlist[1], 16)
		# Gy
		val3 = int(textlist[2], 16) << 8
		val4 = int(textlist[3], 16)
		# Gz
		val5 = int(textlist[4], 16) << 8
		val6 = int(textlist[5], 16)

		val = [val1 | val2, val3 | val4, val5 | val6]

		for i in range(len(val)):
			# if the value is negative
			if (val[i] & 0x8000) == 0x8000:
				val[i] -= 0x10000

		# From 100th of magnetic field to magnetic field
		val = [float(x) / 100.0 for x in val]
		# orientation need to be calcuated in val[3]
		return val


	def decodeHIH6130(self, textlist):
		# Temperature
		val1 = int(textlist[0], 16) << 8
		val2 = int(textlist[1], 16)
		# Humidity
		val3 = int(textlist[2], 16) << 8
		val4 = int(textlist[3], 16)

		val = [val1 | val2, val3 | val4]

		for i in range(len(val)):
			# if the value is negative
			if (val[i] & 0x8000) == 0x8000:
				val[i] -= 0x10000

		val = [x / 100. for x in val]

		return val

	def lightSensors(self, textlist):
		# Raw light
		val1 = int(textlist[0], 16) << 8
		val2 = int(textlist[1], 16)

		val = val1 | val2

		if val == 65535:
			return
		else:
			return val

	def decodeAPDS9006(self, textlist):
		return self.lightSensors(textlist)

	def decodeTSL260RD(self, textlist):
		return self.lightSensors(textlist)

	def decodeTSL250RD(self, textlist):
		return self.lightSensors(textlist)

	def decodeMLX75305(self, textlist):
		return self.lightSensors(textlist)

	def decodeML8511(self, textlist):
		return self.lightSensors(textlist)

	def decodeTMP421(self, textlist):
		# Temperature
		val1 = int(textlist[0], 16) << 8;
		val2 = int(textlist[1], 16)

		val = val1 | val2
		# if the value is negative
		if (val & 0x8000) == 0x8000:
			val -= 0x10000
		val /= 100.
		
		if val > 127:
		# When I2C is not available, return "none"
			return
		else:
			return val